fabric-connector/src/main/java/com/luxoft/fabric/FabricConnector.java

```diff
@@ -254,15 +273,74 @@ public QueryByChaincodeRequest buildQueryRequest(String function, String chainco
         return sendQueryRequest(buildQueryRequest(function, chaincode, message), channelName);
     }
 
+    public CompletableFuture<InvokeResult> invokeEx(String function, String chaincode, byte[]... message) throws Exception {
+        return sendTransactionEx(buildProposalRequest(function, chaincode, message), defaultChannelName);
+    }
+
     public CompletableFuture<BlockEvent.TransactionEvent> invoke(String function, String chaincode, byte[]... message) throws Exception {
-        return invoke(function, chaincode, defaultChannelName, message);
+        return invoke(function, chaincode, defaultChannelName, defaultMaxRetries, message);
     }
 
     public CompletableFuture<BlockEvent.TransactionEvent> invoke(String function, String chaincode, String channelName, byte[]... message) throws Exception {
-        return buildTransactionFuture(buildProposalRequest(function, chaincode, message), channelName);
+        return invoke(function, chaincode, channelName, defaultMaxRetries, message);
     }
 
-    public CompletableFuture<InvokeResult> invokeEx(String function, String chaincode, byte[]... message) throws Exception {
-        return buildTransactionFutureEx(buildProposalRequest(function, chaincode, message), defaultChannelName);
+    public CompletableFuture<BlockEvent.TransactionEvent> invoke(String function, String chaincode, int maxRetries, byte[]... message) throws Exception {
+        return invoke(function, chaincode, defaultChannelName, maxRetries, message);
+    }
+
+    public CompletableFuture<BlockEvent.TransactionEvent> invoke(String function, String chaincode, String channelName, int maxRetries, byte[]... message) throws Exception {
+        CompletableFuture<BlockEvent.TransactionEvent> f = sendTransaction(buildProposalRequest(function, chaincode, message), channelName);
+
+        // Here we handle retry 'maxRetries' times after 'txRetryDelayMs'
+        // Basically we just chain N(='maxRetries') dummy futures that push successful one further
+        // In case of exception it checks transaction error code and either push forward the exception or recreates transaction on retry-able errors
+        for(int i = 0; i < maxRetries; i++) {
+            f = f.thenApply(CompletableFuture::completedFuture)
+                    .exceptionally(t -> {
+                        try {
+                            int validationCode = ((TransactionEventException) t.getCause()).getTransactionEvent().getValidationCode();
+                            if (validationCode == FabricTransaction.TxValidationCode.MVCC_READ_CONFLICT_VALUE ||
```

Comment:
maybe better switch on enum value? like:

```
switch(validationCode) {
   case MVCC_READ_CONFLICT_VALUE:
   .....
   default: 
}
```

Reply:
"Potato, potato" as soon as there are only 2 of around 15 values of enum used here but ok.

Reply:
done